### DISCUSSION TITLE
NewGRF string length limit?

### DISCUSSION THREAD

**Discussion by merni-ns** [OpenTTD/OpenTTD#10418] [Features]
Sparked by #10384.

Many of the UI elements in the game have a limited width. Accordingly, the user is limited in the lengths of strings they can set, like particular vehicle names, vehicle type names, station names, town names, etc. But NewGRFs have no such limit. They can set strings (in particular vehicle type names, station set names and individual station type names) that are extremely long, to the point of either streching the UI window that contains them to huge width, or making parts of the name unreadable in most normal configurations (without resizing the window to make it huge). 

At the same time, in particular for vehicle type names, the user can't edit such long names in a meaningful way. When a name is nearly 100 characters long, the user is unable to either add a few characters or remove a few characters/words. They are restricted to either leaving it as it is or deleting the majority of the name to fit it in 32 characters. This was why I raised #10384 in the first place.

@LC-Zorg pointed out that this is a more fundamental issue than simply increasing the length of names the user can set. It makes no sense for NewGRFs to be able to set such long names that the majority of users will not be able to fully read -- especially for vehicles, which have a descriptive text feature suited exactly for this.

So, how to fix that?

**Option 1**: To increase the length of user-set strings to keep up with the NewGRFs. 

Pros: 
* Easy to implement.
* Existing NewGRFs are unaffected.

Cons:
* Technically NewGRFs could just set even longer names and we would be back in the same issue.
* It doesn't solve the issue of text not fitting on the screen.

**Option 2**: Set a hard limit on NewGRF strings. (Or on some of NewGRF strings where this problem exists.)

Pros:
* Ensures that names are a sane length in all cases.
* User would be able to edit the whole name (although the limit would still have to be increased -- 32 is too small in any case).

Cons:
* Text of existing NewGRFs would be truncated, which might lose valuable content.
  * Generally, for vehicles, NewGRF authors don't duplicate the info in the title, in the descriptive text. Such info might include whether an engine is diesel or electric (which may have gameplay effects), as well as purely informative text.
  * Would it be possible to put the "extra" portion of the name in the descriptive text, at least for vehicles?
  * Impact on other kinds of strings needs to be considered.

**Option 3**: Status quo.
Not desirable. See #10370 and #10384.

Any more options? Have I missed something blindingly obvious? Please comment.

**Comment by PeterN** [comment]
Perhaps we can just change UIs to either crop or wrap text if it's excessively long?

**Reply by merni-ns** [reply]
The vehicle selection UI already crops the excessively long text with a "...". But that doesn't really help either with reading it (you have to stretch it to a big size to see the "(diesel)" at the end) or with editing the names (which truncates them to 32 chars).

**Reply by merni-ns** [reply]
As for wrapping, it might end up looking weird in some places, but sounds like a good idea to try. Again, this doesn't solve the issue with editing.

**Comment by FLHerne** [comment]
**Option 4**: It would be possible to add compile-time warnings about string length to NML and grfcodec to discourage such long strings in some contexts. In principle, NML at least could make it a compile-time error, but I'd be against that.

--

Really I think Option 1 (+ optionally 4) is best.

There are many ways for a grf to do nonsensical things that reduce its usability. You can have invisible vehicles, ten thousand types of vehicle, misaligned sprites that appear on the next track over, articulated vehicles that are longer than any possible station. It's just expected that authors will apply their judgement and make something reasonable.

No-one is forced to use any particular grf, if an author makes bad decisions like this players can choose not to use it or ask for it to be improved. I don't see much point in playing whack-a-mole with individual set design errors.

**Reply by merni-ns** [reply]
I like the idea of Option 1 + 4 + wrapping long strings, myself. Will have to see how that looks in the UI, though, and make sure it doesn't look ugly.

**Reply by FLHerne** [reply]
I feel like wrapping would just encourage authors to use excessively long names instead of better options like the description text.

As far as I know, this problem seems limited to a particular train grf?

IMO we should just treat this like an oversized sprite that gets clipped and causes flickering, it's the same category of problem. (i.e. ignore it in OTTD, and ask the grf author politely if it can be fixed in a future version)

**Reply by merni-ns** [reply]
> As far as I know, this problem seems limited to a particular train grf?

No, there are many grfs that set names of long lengths. The example in the PR is just presumably the worst that could be quickly found.

**Reply by PeterN** [reply]
My wrapping suggestion was mainly for the station name display, rather than the vehicle purchase list.
